## CIS 242

## Spring 2020

## HOMEWORK ASSIGNMENT 4

Please compile your responses use markdown in your Jupyter notebook to answer the questions.  If you prefer, you may also submit a Word or PDF document with the responses along the PDF or HTML version of the completed notebook.   

#### Active notebooks (.ipynb files) or raw code (.py files) will NOT be accepted and no points will be given.

The code part of the files will not be graded, but they will be checked if necessary to verify your findings and recommendations. Point deductions may occur if there are major discrepancies between your written answers and the output from the code. 

Please organize your notebook to have the homework responses at the top and the working code underneath.

Each question is worth 2 points for a total of **24 points**.


## Predicting Loans for Universal Bank

Remember: You were recently hired as a data scientist by Universal Bank. The bank’s Vice President is interested learning what predictive analytics can do for the bank. She supplies you with a dataset containing information on a sample of customers. Here is a description of each variable in the Universal Bank dataset:

* Age: Customer’s age in years
* Experience: Number of years of professional work experience
* Income: Annual income in thousands of dollars (\$000)
* Zip code: Zip code of home address
* Family: Customer’s family size
* CC Avg: Average spending on credit cards per month in thousands of dollars (\$000)
* Education: Education level where 1 = Undergraduate; 2 = Graduate; and 3=Advanced/Professional
* Mortgage: Value of house mortgage if any; in thousands of dollar (\$000)
* Personal.Loan: Did the customer accept a personal loan offered in the bank’s last campaign? 1=Yes; 0 = No.
* Securities.Account: Does the customer have a securities account with the bank? 1 = Yes; 0 = No.
* CD.Account: Does the customer have a certificate of deposit (CD) account with the bank? 1 = Yes; 0 = No.
* Online: Does the customer use Internet banking facilities? 1 = Yes; 0 = No.
* Credit.Card: Does the customer use a credit card issued by Universal Bank? 1 = Yes; 0 = No.


You have already investigated these data and built a logistic regression model to help understand the personal loan market.  Now the VP wants to investigate other predictive models.

#### 1. You are asked to create a NB model to predict personal loans.  What variables will you use?  Why?

We will start of with a Multinomial NB model. For this model, we will use categorical variables as they are most appropriate to use in an Multinomial NB model. These variables are Education, Securities Account, CD Account, Online, and CreditCard.

#### 2. Do you need to make any changes to the data?  If so, what do you change and why?  If not, why not?

No, because all the variables mentioned above are categorical and therefore already expressed as discrete varaibles. This means we do not need to use the ordinal encoder to transform them into catergorical values.

#### 3. Train a NB model.  What set up or parameter choices did you make?  Why?

Model 1.
For the first model, we will use all the binomial/multinomial/categorical variables to see if having more variables in the Multinomial Naive Bayes model increases predictive accuracy. (Variables: Education, Securities Account, CD Account, Online, CreditCard)

In [1]:
import numpy as np
import pandas as pd

UB_df = pd.read_csv('UniversalBank.csv')
bank = pd.DataFrame(UB_df)

#make copy of bank dataframe to manipulate according to need.
bank_copy = bank

from sklearn.naive_bayes import MultinomialNB

In [2]:
#create dataframe with only column values of varaibles mentioned in question 1.
bank_indep1 = bank_copy.drop(columns = ['Age','Experience','Income','ZIP Code','Family','CCAvg','Mortgage','Personal Loan'])
#create separate object with just the personal loan results.
bank_dep = bank.iloc[:, -5]

bank_indep1.head()

Education  Securities Account  CD Account  Online  CreditCard
0          1                   1           0       0           0
1          1                   1           0       0           0
2          1                   0           0       0           0
3          2                   0           0       0           0
4          2                   0           0       0           1

We will be taking 80% of the randomized sample for training and leave 20% for testing as we did in the last assignment.

In [3]:
from sklearn.model_selection import train_test_split

X=bank_indep1
Y=bank_dep

X_train, X_test, Y_train, Y_test = train_test_split(X, Y, test_size = 0.2, random_state=12345)

NB=MultinomialNB()
NB.fit(X_train,Y_train)
#code borrowed from Professor Cannon's CIS 242 Lecture_4.pdf.

MultinomialNB(alpha=1.0, class_prior=None, fit_prior=True)

#### 4. Test the accuracy of your model.  What is the overall accuracy of the model? What did it do well?  What did it not do well?


In [4]:
from sklearn.metrics import accuracy_score
nb_pred = NB.predict(X_test)

print('Model accuracy:', accuracy_score(Y_test, nb_pred))
pd.crosstab(nb_pred, Y_test, rownames = ["Predicted"], colnames = ["Actual"])
#code borrowed from Professor Cannon's CIS 242 Lecture_4.pdf.

Model accuracy: 0.904


Actual       0   1
Predicted         
0          904  96

The overall accuracy of this Naive Bayes model is 90.4%

#### 5. Train at least two more versions of the model.  What changes did you make?  Why?  What was the outcome?

Model 2.
For the second model, we will use the GaussianNB to take in continuous varaibles that had positive coefficients in relation to Personal Loan last assignment. (Variables: Experience, Income, CCAvg)

In [5]:
bank_indep2 = bank_copy.drop(columns = ['Age','Education','ZIP Code','Personal Loan','Securities Account','Online',
                                        'CreditCard','Family','Mortgage','CD Account'])
bank_indep2.head()

Experience  Income  CCAvg
0           1      49    1.6
1          19      34    1.5
2          15      11    1.0
3           9     100    2.7
4           8      45    1.0

In [6]:
X2=bank_indep2
Y2=bank_dep

from sklearn.naive_bayes import GaussianNB
gnb = GaussianNB()

X2_train, X2_test, Y2_train, Y2_test = train_test_split(X2, Y2, test_size = 0.2, random_state=12345)
gnb.fit(X2_train,Y2_train)
gnb_pred = gnb.predict(X2_test)

print('Model accuracy:', accuracy_score(Y2_test, gnb_pred))
pd.crosstab(gnb_pred, Y2_test, rownames = ["Predicted"], colnames = ["Actual"])

Model accuracy: 0.886


Actual       0   1
Predicted         
0          830  40
1           74  56

The overall accuracy of this Gaussian Naive Bayes model is 88.6%

Model 3. 
In this model, we will only use continuous variables with the Gaussian Naive Bayes model to compare the model's accuracy with model 1. (Variables: Age, Experience, Income, CCAvg, Mortgage)

In [7]:
bank_indep3 = bank_copy.drop(columns = ['Education','ZIP Code','Family','Personal Loan','Securities Account','CD Account',
                                       'Online','CreditCard'])
bank_indep3.head()

Age  Experience  Income  CCAvg  Mortgage
0   25           1      49    1.6         0
1   45          19      34    1.5         0
2   39          15      11    1.0         0
3   35           9     100    2.7         0
4   35           8      45    1.0         0

In [8]:
X3=bank_indep3
Y3=bank_dep

X3_train, X3_test, Y3_train, Y3_test = train_test_split(X3, Y3, test_size = 0.2, random_state=12345)
gnb.fit(X3_train,Y3_train)
gnb_pred2 = gnb.predict(X3_test)

print('Model accuracy:', accuracy_score(Y3_test, gnb_pred2))
pd.crosstab(gnb_pred2, Y3_test, rownames = ["Predicted"], colnames = ["Actual"])

Model accuracy: 0.883


Actual       0   1
Predicted         
0          828  41
1           76  55

The overall accuracy of this Gaussian Naive Bayes model is 88.3%

It turns out with this data set, model 1 shows the highest overall predictive accuracy.

#### 6. Now you are asked to create a KNN model to predict personal loans.  What variables will you use?  Why?

We will use the same varaibles as NB model 2, as none of the variables in model 2 are categorical and it has a higher accuracy than model 3. We discussed in class that the KNN model does not work with categorical variables if they are not represented as numerical data types, so we will only be using the ones in model 2. (Variables: Experience, Income, CCAvg)

#### 7. Do you need to make any changes to the data?  How is this similar or different to the NB model?  Why?

We need to standardize the data since distance calculations are affected by different scale of variables, similar to having to standardize data for the NB model.

In [9]:
from sklearn.preprocessing import MinMaxScaler
X2= MinMaxScaler().fit_transform(X2)

#### 8. Train a KNN model. What set up or parameter choices did you make? Why?

KNN model 1. Chose 2 for n_neighbors parameter because there are only two outcomes for 'Personal Loan' and chose 'Euclidian' as metric parameter because it's the most widely used method when calculating 2-dimensional distance between two points.

In [10]:
from sklearn.neighbors import KNeighborsClassifier

np.random.seed(123)
train = X2_train
test = X2_test
pred_knn = KNeighborsClassifier(n_neighbors = 2, metric = "euclidean")
pred_knn.fit(X2_train, Y2_train)
#code borrowed from Professor Cannon's CIS 242 Lecture_4.pdf.

KNeighborsClassifier(algorithm='auto', leaf_size=30, metric='euclidean',
                     metric_params=None, n_jobs=None, n_neighbors=2, p=2,
                     weights='uniform')

#### 9. Test the accuracy of your model. What is the overall accuracy of the model? What did it do well? What did it not do well?

In [11]:
knn_pred = pred_knn.predict(X2_test)

print('Model accuracy:', accuracy_score(Y2_test, knn_pred))
pd.crosstab(knn_pred, Y2_test, rownames = ["Predicted"], colnames = ["Actual"])

Model accuracy: 0.909


Actual       0   1
Predicted         
0          886  73
1           18  23

Overall accuracy of this KNN model 1 comes out to be 90.9%. Accuracy came out to be quite high so the model performed well, but it might've been affected by outliers as KNN models are known to be susceptible to. We know there are no missing values in the Universal Bank dataset so that's not an issue.

#### 10. What changes can you make that might affect the KNN model?  Train at least two more models.   What is the result?

KNN Model 2. Using only NB model 1 variables this time. These varaibles are categorical but represented in numeric data types so the KNN model can compute them.

In [12]:
train = X_train
test = X_test
pred_knn = KNeighborsClassifier(n_neighbors = 2, metric = "euclidean")
pred_knn.fit(X_train, Y_train)

KNeighborsClassifier(algorithm='auto', leaf_size=30, metric='euclidean',
                     metric_params=None, n_jobs=None, n_neighbors=2, p=2,
                     weights='uniform')

In [13]:
import matplotlib.pyplot as plt
plt.scatter(X['CreditCard'],Y)
plt.xlabel('Credit Card, 1=yes 0=no')
plt.ylabel('Personal Loan, 1=yes 0=no')
plt.show()
#this is just an example to show if there are any noticeable patterns between one of the variables,'CreditCard', and 'Personal Loan'.

<Figure size 640x480 with 1 Axes>

In [14]:
knn_pred = pred_knn.predict(X_test)

print('Model accuracy:', accuracy_score(Y_test, knn_pred))
pd.crosstab(knn_pred, Y_test, rownames = ["Predicted"], colnames = ["Actual"])

Model accuracy: 0.916


Actual       0   1
Predicted         
0          900  80
1            4  16

Overall accuracy of this KNN model 2 comes out to be 91.6%.

KNN model 3. For this model, we will use the variables from NB model 3, which were all continuous variables. (Variables: Age, Experience, Income, CCAvg, Mortgage)

In [15]:
X3= MinMaxScaler().fit_transform(X3)
#scale standardization is necessary as in KNN model 1 because the scales of variables are vastly different.

In [16]:
train = X3_train
test = X3_test
pred_knn = KNeighborsClassifier(n_neighbors = 2, metric = "euclidean")
pred_knn.fit(X3_train, Y3_train)

KNeighborsClassifier(algorithm='auto', leaf_size=30, metric='euclidean',
                     metric_params=None, n_jobs=None, n_neighbors=2, p=2,
                     weights='uniform')

In [17]:
knn_pred = pred_knn.predict(X3_test)

print('Model accuracy:', accuracy_score(Y3_test, knn_pred))
pd.crosstab(knn_pred, Y3_test, rownames = ["Predicted"], colnames = ["Actual"])

Model accuracy: 0.904


Actual       0   1
Predicted         
0          887  79
1           17  17

Overall accuracy of this KNN model 2 comes out to be 90.4%.

#### 11. What are the relative strengths and weaknesses of the various models?  Mention at least 3 points. 

Strengths of NB model: 
1. It's generative compared to KNN so it handles missing values much better.
2. Despite the independence assumption being rarely true, the model performs very well.
3. Because it's a non-lazy supervised training model, it requires less training data set.

Weaknesses of NB model:
1. The independence assumption among predictors very rarely holds true in real life.
2. If a categorical variable that was not introduced in the training set is introduced in the test set, it is automatically assigned 0 probabilty and the model will be unable to make predictions with it.

Strenghts of KNN model:
1. No training period is required, so adding new data is seamless.
2. The algorithm is simple and easy to implement due to few parameters.
3. Makes no assumptions that are mostly untrue in real life.

Weaknesses of KNN model:
1. Always have to convert categorical varaibles into dummy/numeric type variables.
2. Scaling is crucial since it measures distances and not scaling leads to wrong predictions.
3. Sensitive to outliers and missing values.

#### 12. What would you recommend to the VP?  Why? Remember, you have 3 different models (and their variants) to choose from.  

I would recommend using the KNN model with categorical variables that were used in KNN model 2 and NB model 1. Normally it would make little sense to use categorical variables for a KNN model but fortunately in this dataset, all teh categorical varaibles are represented as dummy/numeric data types so the KNN model can actually run its algorithm and produce an overall accuracy of 91.6%. With this model, we can use Education, Securities Account, CD Account, Online, and CreditCard varaibles to predict whether a customer will take out a personal loan or not more than 9 out of 10 times.

## Final project group members:  with whom are you working on your final project?
Please make sure one member of your group posts the names of people in your group to [this Google doc](https://docs.google.com/spreadsheets/d/15s2RIzZ7SqFMqi5ZAsUdmTrboNrQI2TKb0P2AbNpU9Q/edit?usp=sharing) before class on **February 13**.  If you have fewer than 4 members, please indicate if you would be open to an additional collaborators by using the terms "Open" or "Closed" to describe your group. 

If you have not yet found/formed a group, look for groups marked "Open" to see about joining them.  Or post that you are a one person "Open" group so that others can find you. 

## Quick survey: 
Please take a minute to provide feedback on how you think the course is going so far [using this link](https://docs.google.com/forms/d/16XmbU7PtVFAqv3aOGPwBIkmOD_IeMGVELde4ZOmCyeI/edit)

The responses are anonymous and will be helpful for me to understand how I can try to make sure you get the most out of the class. 

### If we get a response rate of 75% or above, everyone gets an extra point on this assignment!  That means that 75% of the class provides honest feedback and not 75% of the class says nice things.  :-)

## Working code below

All work is shown underneath each question.